### PR TITLE
validating the required headers while uploading points data

### PR DESCRIPTION
**Pull request by emre2038**
## Description
handling the errors that occurs when points data upload is missing required columns. assumed required columns are : 
* name
* longitude
* latitude

## Related Issue
https://wazimap.atlassian.net/browse/WNCM-274

## How to test it locally
* navigate to admin panel
* try to create a collection by uploading a file that does not contain any of the columns above
* confirm that you get a validation error

image

### Added
* `tests/points/admin/test_category_admin.py` for automated tests

### Updated
* `wazimap_ng/points/admin/forms/category_admin_form.py` to add the `clean` method for validations

## Checklist

- [x]  🚀 is the code ready to be merged and go live?
- [x]  🛠 does it work (build) locally

### Pull Request

- [x]  📰 good title
- [x]  📝good description
- [x]  🔖 issue linked
- [x]  📖 changelog filled out

### Commits

- [x]  commits are clean
- [x]  commit messages are clean

### Code Quality

- [x]  🚧 no commented out code
- [x]  🖨 no unnecessary logging
- [x]  🎱 no magic numbers
- [ ]  black was run locally (as part of the pre-commit hook)

### Testing

- [x]  ✅ added (appropriate) unit tests
- [x]  💢 edge cases in tests were considered
- [x]  ✅ ran tests locally & are passing

### CODE REVIEW

File: wazimap_ng/points/admin/forms/category_admin_form.py [hunk 1]
```diff
@@ -22,6 +27,23 @@ def __init__(self, *args, **kwargs):
...
 
+    def clean(self):
+        cleaned_data = super(CategoryAdminForm, self).clean()
+        document = cleaned_data.get('import_collection', None)
+        required_headers = ["name", "longitude", "latitude", "id", "status", "status last updated", "class", "volume",
+                            "discharges into"]
+        if document is not None:
+            headers = pd.read_csv(BytesIO(document.read()), nrows=1, dtype=str).columns.str.lower().str.strip()
+            missing_headers = [
+                h.capitalize() for h in list(set(required_headers) - set(headers))
```
Comment (michaelglenister) [inline wazimap_ng/points/admin/forms/category_admin_form.py:38]: So this is where the magic happens. Removing headers provided from a list of required headers and then check if there are headers remaining from required_headers. Nice

File: tests/points/admin/test_category_admin.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+
+@pytest.mark.django_db
+class TestCategoryAdmin:
+    def create_csv_file(self, data):
+        csvfile = StringIO()
+        csv.writer(csvfile).writerows(data)
+        return csvfile.getvalue()
+
+    def test_uploading_file_with_missing_headers(self, client, superuser):
```
Comment (michaelglenister) [inline tests/points/admin/test_category_admin.py:16]: Happy with the test, I don't think checking one missing header is enough

**Review by michaelglenister — APPROVED** [review]
Changes are good to go, thanks

File: wazimap_ng/points/admin/forms/category_admin_form.py [hunk 3]
```diff
@@ -22,6 +27,23 @@ def __init__(self, *args, **kwargs):
                 self.fields["description"].initial = metadata.description
                 self.fields["licence"].initial = metadata.licence
 
+    def clean(self):
+        cleaned_data = super(CategoryAdminForm, self).clean()
+        document = cleaned_data.get('import_collection', None)
+        required_headers = ["name", "longitude", "latitude", "id", "status", "status last updated", "class", "volume",
```
Comment (goyal1092) [inline wazimap_ng/points/admin/forms/category_admin_form.py:33]: Hey @emre2038 
There are only 3 headers that are required : name, longitude & latitude

**Review by goyal1092 — CHANGES_REQUESTED** [review]
Can you please also remove the clean method in the models.py 
https://github.com/OpenUpSA/wazimap-ng/blob/staging/wazimap_ng/points/models.py#L107

File: tests/points/admin/test_category_admin.py (hunk 2 above)
Comment (emre2038) [reply tests/points/admin/test_category_admin.py]: added more tests to confirm that the upload fails for each of the missing required headers(name, longitude, latitude) and it does not fail when a non-required column in missing

**Review by goyal1092 — APPROVED** [review]
LGTM 👍

**Comment by goyal1092** [comment]
Review App : https://pr-399.herokuapp.com/
